ggshield/cmd/iac/scan/all.py

```diff
@@ -0,0 +1,95 @@
+from pathlib import Path
+from typing import Any, Optional, Sequence, Union
+
+import click
+from pygitguardian.iac_models import IaCScanParameters, IaCScanResult
+
+from ggshield.cmd.iac.scan.iac_scan_common_options import (
+    add_iac_scan_common_options,
+    directory_argument,
+    update_context,
+)
+from ggshield.cmd.iac.scan.iac_scan_utils import (
+    IaCSkipScanResult,
+    create_output_handler,
+    handle_scan_error,
+)
+from ggshield.iac.collection.iac_path_scan_collection import IaCPathScanCollection
+from ggshield.iac.filter import get_iac_files_from_path
+from ggshield.scan import ScanContext, ScanMode
+
+
+# Changes to arguments must be propagated to default_command
+@click.command()
+@add_iac_scan_common_options()
+@directory_argument
+@click.pass_context
+def scan_all_cmd(
+    ctx: click.Context,
+    exit_zero: bool,
+    minimum_severity: str,
+    ignore_policies: Sequence[str],
+    ignore_paths: Sequence[str],
+    directory: Optional[Path] = None,
+    **kwargs: Any,
+) -> int:
+    """
+    Scan a directory for all IaC vulnerabilities in the current state.
+    """
+    if directory is None:
+        directory = Path().resolve()
+    update_context(ctx, exit_zero, minimum_severity, ignore_policies, ignore_paths)
+
+    result = iac_scan_all(ctx, directory)
+    return display_iac_scan_all_result(ctx, directory, result)
+
+
+def iac_scan_all(
+    ctx: click.Context, directory: Path
+) -> Union[IaCScanResult, IaCSkipScanResult, None]:
+    paths = get_iac_files_from_path(
```

Comment:
`get_iac_files_from_path` generates a list of path using the `get_filepaths` function relying on a set.

This cause the path list to have inconsistent order and therefore the tar generated have a different byte order when called twice on the same directory. This cause the hash_key of GIM cache to be different when scanning twice the same directory so the cache is ignored.

If this is the case, a path.sort() might be necessary here.